### PR TITLE
Implicit withdrawals for corrected AS-paths

### LINKED ISSUE #74: Investigate how we can automatically mark implicit withdrawals on hijacks with corrected paths
**Is your feature request related to a problem? Please describe.**
Currently we mark a hijack as withdrawn only upon receiving explicit withdrawals. However,
hijacked paths can be implicitly withdrawn (e.g., by being changed to a legal path).

**Describe the solution you'd like**
Investigate how we can treat path changes as implicit withdrawals (we can use the currently available fields, no need for db changes). For example, if a new **benign** BGP update comes, and it includes a path to a prefix P and is learned from monitor X, and P and X are the prefix and one of the monitors of an ongoing hijack, then we can mark X as having seen a withdrawal.

*Optional, to discuss:* Note that if the benign update is for a sub-prefix of a currently hijacked prefix, then we need to see all sub-prefixes that build the hijacked super-prefix being corrected. to consider a monitor withdrawal.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by vkotronis**
### Description of PR

What component(s) does this PR affect?

- [X] Back-End (Database, Microservices, Containers, etc)
- [ ] Front-End (Flask, API, etc)
- [ ] Docs
- [ ] Build System

Does the PR require changes on other components? If yes, please mark the components:

- [ ] Back-End (Database, Microservices, Containers, etc)
- [ ] Front-End (Flask, API, etc)
- [X] Docs
- [ ] Build System

### Related Issue

Resolves #74

### Solution

Screenshot_2019-06-05 ARTEMIS - BGP Updates

Screenshot_2019-06-05 ARTEMIS - Hijack

Screenshot_2019-06-06 ARTEMIS - BGP Updates(1)

### Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Docs update
- [ ] None of the above

### Checklist:

- [X] I have read the **[contributing guide](https://github.com/FORTH-ICS-INSPIRE/artemis/blob/master/CONTRIBUTING.md)** and my code conforms to the guidelines.
- [X] This change requires a change in the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by vkotronis** [comment]
Only issue: "healed" BGP paths leading to sub-prefixes of the hijacked prefix (i.e., monitoring of deaggregation effects on hijack mitigation)

File: backend/core/utils/__init__.py [hunk 1]
```diff
@@ -226,6 +226,11 @@ def purge_redis_eph_pers_keys(redis_instance, ephemeral_key, persistent_key):
     redis_pipeline.delete(ephemeral_key)
     redis_pipeline.srem("persistent-keys", persistent_key)
     redis_pipeline.delete("hij_orig_neighb_{}".format(ephemeral_key))
+    for element in redis_instance.sscan_iter(
+        "hij_prefix_peer_{}".format(ephemeral_key)
+    ):
+        redis_pipeline.srem("hij_prefix_peer_all", element)
```
Comment (vkotronis) [inline backend/core/utils/__init__.py]: TODO: make sure that this combination is not present in any other ongoing hijack for the same prefix (of different type) with this peer. However, this is extremely rare. I have a way to deal with this by inputting (in redis) sets of hijack keys with the name: hij_prefix_<prefix>_peer_<peer>. Will fix soon, should not be very hard.

File: backend/core/detection.py [hunk 2]
```diff
@@ -830,6 +844,33 @@ def mark_outdated(self, hij_key: str, redis_hij_key: str) -> NoReturn:
...
+            """
+            Checks if a benign BGP update should trigger an implicit withdrawal
+            """
+            # log.debug('{}'.format(monitor_event['key']))
+            prefix = monitor_event["prefix"]
+            peer_asn = monitor_event["peer_asn"]
+            if self.redis.exists("prefix_{}_peer_{}_hijacks".format(prefix, peer_asn)):
+                # generate implicit withdrawal
+                withdraw_msg = {
+                    "service": "implicit-withdrawal",
```
Comment (slowr) [inline backend/core/detection.py]: I really really don't like this hacky approach of sending it as a monitor :/

File: backend/core/detection.py [hunk 3]
```diff
@@ -830,6 +844,33 @@ def mark_outdated(self, hij_key: str, redis_hij_key: str) -> NoReturn:
...
+                    "type": "W",
+                    "prefix": prefix,
+                    "path": [],
+                    "orig_path": {"triggering_bgp_update": monitor_event},
+                    "communities": [],
+                    "timestamp": monitor_event["timestamp"],
+                    "peer_asn": peer_asn,
+                }
+                key_generator(withdraw_msg)
+                self.producer.publish(
```
Comment (slowr) [inline backend/core/detection.py]: why do we need to send to the update exchange and not just send the correct messages to the withdraw exchange directly?

Comment (vkotronis) [reply backend/core/detection.py]: are these two equivalent, regarding their handling? If yes, I will use the withdraw exchange, no problem.

File: backend/core/detection.py (hunk 2 above)
Comment (vkotronis) [reply backend/core/detection.py]: It is not really a hack, but code reuse :) ; we need to somehow trigger implicit withdrawals when the paths change. I do not see any other good way of doing this without changing large chunks of code (the current solution takes advantage of what ARTEMIS does well, which is processing updates quickly, without requiring changes in the logic, besides the pseudo-withdrawal messages). However, any other ideas on how to achieve the same end result are welcome ;) .

File: backend/core/detection.py (hunk 3 above)
Comment (vkotronis) [reply backend/core/detection.py]: also as I see in the rest of the detection code:
```
elif monitor_event["type"] == "W":
                self.producer.publish(
                    {
                        "prefix": monitor_event["prefix"],
                        "peer_asn": monitor_event["peer_asn"],
                        "timestamp": monitor_event["timestamp"],
                        "key": monitor_event["key"],
                    },
                    exchange=self.update_exchange,
                    routing_key="withdraw",
                    priority=0,
)
```
i.e., we use the update exchange to send withdrawals. This is yet another withdrawal, so it should go over the same pipeline, correct?

Comment (slowr) [reply backend/core/detection.py]: withdraw routing_key goes directly to the database module while the update routing_key requeues it to the detection module. Does it make sense to go through the detection again?

Comment (vkotronis) [reply backend/core/detection.py]: if there are no implications I will send it directly, good call.

Comment (vkotronis) [reply backend/core/detection.py]: see slack discussion, sending it directly results to non-registration of the withdrawal to the db and the loss of useful information.

Comment (vkotronis) [reply backend/core/detection.py]: resolving as discussed. The implicit withdrawal needs to be handled by the db, and then the detector should be notified and synced with the DB w.r.t. this as with the other withdrawals (current solution, compliant with the initial design logic).

**Pull request merged**